src/storage.rs

```diff
@@ -157,101 +392,411 @@ impl Display for StreamName {
+    /// completion, not by submission; use `BatchAppendResult::batch_index`
+    /// to correlate with the original input order.
+    pub async fn append_table_batches_concurrent<M>(
+        &self,
+        table_batches: Vec<TableBatch<M>>,
+        max_concurrent_streams: usize,
+        trace_id: &str,
+    ) -> Result<Vec<BatchAppendResult>, BQError>
+    where
+        M: Message + Send + 'static,
+    {
+        if table_batches.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        let batches_num = table_batches.len();
+        let semaphore = Arc::new(Semaphore::new(max_concurrent_streams));
+
+        let mut join_set = JoinSet::new();
+        for (idx, table_batch) in table_batches.into_iter().enumerate() {
+            // Acquire a concurrency slot and hold it until responses are fully drained.
+            let permit = semaphore.clone().acquire_owned().await?;
+
+            let stream_name = table_batch.stream_name.clone();
+            let table_descriptor = table_batch.table_descriptor;
+            let rows = table_batch.rows;
+            let trace_id = trace_id.to_string();
+            let client = self.clone();
+
+            join_set.spawn(async move {
+                // We compute the proto schema once for the entire batch. We might want to compute it
+                // once per stream but for now this is fine.
+                let proto_schema = Self::create_proto_schema(&table_descriptor);
+
+                // Create an atomic counter for tracking bytes sent for this batch.
+                let bytes_sent_counter = Arc::new(AtomicUsize::new(0));
+
+                // Build the request stream which will split the request into multiple requests if
+                // necessary.
+                let request_stream =
+                    AppendRequestsStream::new(rows, proto_schema, stream_name, trace_id, bytes_sent_counter.clone());
+
+                let mut batch_responses = Vec::new();
+
+                // Make the request for append rows and poll the response stream until exhausted. Since
+                // we might send multiple requests over the same stream, we will also receive multiple
+                // responses.
+                match Self::new_authorized_request(client.auth.clone(), request_stream).await {
+                    Ok(request) => match client.connection_pool.get_client().await {
+                        Ok(write_client) => {
+                            // We clone the client to cheaply issue multiple requests over the same connection
+                            // without holding onto the original connection object.
+                            //
+                            // This approach utilizes connections efficiently, assuming each call to the pool
+                            // returns a different connection — which holds true when using `Fifo` queuing.
+                            let mut client = write_client.deref().clone();
```

Comment:
This is a trick that I have used to cycle through connections and leverage the multiplexing capabilities of the connection. If we don't do this, the connection will be held for the entire duration of the streaming, so we won't leverage multiplexing and just use multiple connections.